lib/ansible/modules/cloud/docker/docker_swarm_service.py

```diff
@@ -1125,6 +1096,11 @@ def run(self):
         return msg, changed, rebuilt, differences.get_legacy_docker_diffs(), facts
 
 
+def _detect_publish_mode_usage(client):
+    for publish_def in client.module.params.get('publish', []):
+        return 'mode' in publish_def.keys()
```

Comment:
So `publish_def['mode'] == None` is also problematic?

Also, as I understand it from the code which interprets `publish`:
```.py
        ports = {}
        for port in self.publish:
            if port['mode']:
                ports[int(port['published_port'])] = (int(port['target_port']), port['protocol'], port['mode'])
            else:
                ports[int(port['published_port'])] = (int(port['target_port']), port['protocol'])
```
`mode` always **must** be there. So if `'mode'` is not in `publish_def.keys()`, the parsing code (above) will crash. Or did I miss something?

Reply:
Good catch. I think that this is just a bug. I'm changing that code to:

```python
if 'mode' in port:
    ports[int(port['published_port'])] = (int(port['target_port']), port['protocol'], port['mode'])
else:
    ports[int(port['published_port'])] = (int(port['target_port']), port['protocol'])
```

Reply:
Hmm, this probably changes the behavior of the module in the special case that `mode` in port`, but `mode['port']` is falsish. I think it would be better to assume that `mode` can be there but falsish, though (i.e. use `if port.get('mode'):`); the advantage is that we can use `suboptions` to describe the form of the `dict`s in the `published` list (in arg_spec), which will result in Ansible making sure that every `dict` in the list contains all fields it knows of (and none more). Then `mode` will always be in `port`, but it will be `None` if the user didn't specify it.

Reply:
Sounds reasonable, I've pushed the change to use `if port.get('mode'):`.